tutorials/shading/your_first_shader/what_are_shaders.rst

```diff
@@ -0,0 +1,111 @@
+.. _doc_what_are_shaders:
+
+What are shaders?
+=================
+
+Introduction
+------------
+
+So, you have decided to give shaders a try. You have likely heard that they can be used to 
+create interesting effects that run incredibly fast. You have also likely heard that they
+are terrifying. Both are true. 
+
+Shaders can be used to create a wide range of effects (in fact everything drawn in a modern
+rendering engine is done with shaders). 
+
+Writing shaders can also be very difficult for people unfamiliar with them. Godot tries to make writing
+shaders a little easier by exposing many useful built-in features and handling some of the
+lower-level initialization work for you. However, GLSL (the OpenGL Shading Language, which Godot uses)
+is still unintuitive and restricting, especially for users who are used to GDScript.
+
+But what are they?
+------------------
+
+Shaders are a special kind of program that runs on Graphics Processing Units (GPUs). All computers
+have some sort of GPU, either one integrated into their CPU or discrete (meaning it is a separate 
+hardware component). GPUs are especially useful for rendering because they are optimized for running
+thousands of instructions at once.
+
+The output of the shader is the colored pixels of the object drawn to the viewport. Shaders operate
+with a fixed-functionality pipeline. The standard process is the vertex -> fragment shader pipeline. The vertex
+shader is used to decided where each vertex (point in a 3D model, or corner of a Sprite) goes and the 
+fragment shader decides what color individual pixels receive. 
+
+Suppose you want to update all the pixels in a texture to a given color, on the CPU you would write:
+
+:: 
+
+  for x in range(width):
+    for y in range(height):
+      set_color(x, y, some_color)
+
+In a shader you are given access only to the inside of the loop so what you write looks like this:
+
+.. code-block:: glsl
+
+  // function called for each pixel
+  void fragment() {
+    COLOR = some_color;
+  }
+
+You have no control over how this function is called. So you have to design your shaders
+differently from how you would design programs on the CPU.
+
+A consequence of this fixed-funtionality is that you cannot access the results from a previous 
```

Comment:
I think this actually confused what *fixed-functionality* is in a graphics context.

A fixed function pipeline is one where there aren't any shaders that the GPU executes; the GPU state is instead manipulated directly via an API, instead of executing a shader.

>The term Fixed Function Pipeline commonly refers to the set of configurable processing state present in older versions of OpenGL that have since been replaced by Shaders of some sort. While the current OpenGL pipeline still retains non-programmable state, that is not typically what people are referring to when they say "fixed function".
https://www.khronos.org/opengl/wiki/Fixed_Function_Pipeline